src/odemis/driver/xt_client.py

```diff
@@ -2674,14 +2675,16 @@ def _acquire(self):
                 self._acq_wait_start()
                 logging.debug("Preparing acquisition")
                 while True:
-                    if self._acq_should_stop():
+                    logging.debug("Start acquiring an image")
+                    if self._acq_should_stop(timeout=0.5):
```

Comment:
@pieleric this timeout was still needed, otherwise the GEN_STOP message was still too soon. This will slow down the live stream a bit, do you have any idea for a better fix?

Reply:
I think you mean that the GEN_STOP would still be sent after the `scan_image()` has been called (so more like "too late"), right? I guess what happens is that the first image is acquired, it's sent to the client (ie, the GUI), then the client unsubscribes, which calls stop_generate() and sends a GEN_STOP message. However, in the meantime, the next acquisition has already started. Then `stop_generate()` should cancel the acquisition, but maybe that doesn't work well? So this helps by waiting a little bit before starting the next acquisition, to "makes sure it's worthy starting the next acquisition". 

0.5 s sounds a lot of time (for the client to receive the image and decide to unsubscribe). Wouldn't 0.1 s be enough?

That said, all the behaviour so far is how the the detectors are supposed to behave: if you subscribe to them, they try to acquire as fast as possible. If you are not interested anymore, then you unsubscribe, and the last acquisition is cancelled. Of course here, as you keep acquiring-cancelling, it's at least inefficient, and maybe the current cancellation behaviour is still not proper (do you see a "scan image aborted" every time?).

I think the correct way to handle it using the Odemis API would be to use software trigger (aka "event"). This tells the detector that before starting a new acquisition, it should wait for a trigger. You'd need to extend the acquisition thread to support it (see andorcam2, andorcam3, or semcomedi for examples).

1. the client indicates that the detector should be "synchronized on" an event. (-> nothing happens yet)
2. the client subscribes to the detector -> the detector "gets ready" (not really much happening in your case here)
3. the client sends an event (aka software trigger) -> the acquisition starts
4. the acquisition thread receives an image and waits for another event (-> the client receives the image)
5. the client moves the stage, do more fancy stuff
6. repeat from step 3 until all tiles are acquired.
7. the client unsubscribe
8. the client "synchronize on" nothing.

Reply:
> I think you mean that the GEN_STOP would still be sent after the scan_image() has been called (so more like "too late"), right? I guess what happens is that the first image is acquired, it's sent to the client (ie, the GUI), then the client unsubscribes, which calls stop_generate() and sends a GEN_STOP message. 

you are correct, I meant too late.

> However, in the meantime, the next acquisition has already started. Then stop_generate() should cancel the acquisition, but maybe that doesn't work well? So this helps by waiting a little bit before starting the next acquisition, to "makes sure it's worthy starting the next acquisition".

Indeed I think it's not possible to actually stop scanning, which the code in `stop_generate` was supposed to do. Therefore once `scan_image` has started it will finish a full scan.

> 0.5 s sounds a lot of time (for the client to receive the image and decide to unsubscribe). Wouldn't 0.1 s be enough?

I checked the logs and usually GEN_STOP is usually around 0.17 s after `logging.debug("Start acquiring an image")`, so 0.2 is probably safe.